lib/internal/per_context/safethenable.js

```diff
@@ -0,0 +1,71 @@
+'use strict';
+
+const {
+  PromisePrototypeCatch,
+  PromisePrototypeThen,
+  ReflectApply,
+  SafeWeakMap,
+} = primordials;
+
+const { isPromise } = require('internal/util/types');
+
+const cache = new SafeWeakMap();
+
+// `SafeThenable` should be used when dealing with user provided `Promise`-like
+// instances. It provides two methods `.then` and `.catch`.
+// Wrapping uses of `SafeThenable` in `try`/`catch` may be useful if the
+// accessing `.then` property of the provided object throws.
+class SafeThenable {
+  #is_promise;
+  #makeUnsafeCalls;
+  #target;
+  #cachedThen;
+  #hasAlreadyAccessedThen;
+
+  constructor(thenable, makeUnsafeCalls = false) {
```

Comment:
A comment here describing the intent of `makeUnsafeCalls` would be good.

Also, I generally prefer avoiding boolean arguments in favor of named flags or an options object. That is, either `new SafeThenable(promise, kMakeUnsafeCalls)` where `kMakeUnsafeCalls === 1` (allows additional flags to be added later using xor) or `new SafeThenable(promise, { unsafeCalls: true })`.